### PR TITLE
Conformal prediction intervals

### PR DESCRIPTION
**Pull request by topepo**
Produces prediction intervals based on [https://doi.org/10.1080/01621459.2017.1307116](https://doi.org/10.1080/01621459.2017.1307116) (algorithm 1): 

image

The default for finding the intervals uses a fast root finding approach rather than grid search (there is an option to do that though). This method is 2-5-fold faster and seems to have acceptable properties (simulations to check coverage are forthcoming).

Currrently only implemented for regression models

### CODE REVIEW

File: R/conformal_infer.R [hunk 1]
```diff
@@ -0,0 +1,455 @@
...
+#' @param required_pkgs An optional character string for which packages are
+#' required.
+#' @param seed A single integer used to control randomness when models are
+#' (re)fit.
+#' @return A list object with the options given by the user.
+#' @export
+control_conformal_infer <-
+  function(method = "search", trial_points = 100, var_multiplier = 10,
+           max_iter = 100, tolerance = .Machine$double.eps^0.25, progress = FALSE,
+           required_pkgs = character(0), seed = NULL) {
```
Comment (simonpcouch) [inline R/conformal_infer.R:439]: Can we just set a default `seed = sample.int(10^5, 1)` and avoid the `is.null` conditional?

File: R/conformal_infer.R [hunk 2]
```diff
@@ -0,0 +1,455 @@
...
+
+  dplyr::tibble(
+    .pred_lower = get_root(lower, ctrl),
+    .pred_upper = get_root(upper, ctrl)
+  )
+}
+
+get_root <- function(x, ctrl) {
+  if (inherits(x, "try-error")) {
+    rlang::warn("Could not finish the search process due to errors.")
```
Comment (simonpcouch) [inline R/conformal_infer.R:405]: This is another place where I'd wonder if it's helpful for us to pass along the condition message.

File: R/conformal_infer.R [hunk 3]
```diff
@@ -0,0 +1,455 @@
...
+#' @details
+#' This function implements what is usually called "full conformal inference"
+#' (see Algorithm 1 in Lei _et al_ (2018)) since it uses the entire training
+#' set to compute the intervals.
+#'
+#' This function prepares the objects for the computations. The [predict()]
+#' method computes the intervals for new data.
+#'
+#' For a given prediction, different "trial" values of the outcome are evaluated
+#' but augmenting the training set with the sample being predicted (with the
```
[the remark this answers is not in the document]
Comment (simonpcouch) [reply R/conformal_infer.R:27]: Now that I have a better sense for what this algorithm is doing, "augment" gives the wrong impression for me. Maybe that's the broom-brain speaking. :)

How about:

> For a given `new_data` observation, the predictors are appended to the original training set. Then, different "trial" values of the outcome are substituted in for that observation's outcome and the model is re-fit.

File: R/conformal_infer.R [hunk 4]
```diff
@@ -0,0 +1,455 @@
...
+#' trial value as the outcome). From this model, the residual associated with
+#' the trial value is compared to a quantile of the distribution of the other
+#' residuals. Usually the absolute values of the residuals are used. Once the
```
Comment (simonpcouch) [inline R/conformal_infer.R:30]: Was originally unclear to me whether the "other residuals" were the residuals associated with the original training data for a given observation or the residuals associated with the other trial values across models.

**Review by simonpcouch — COMMENTED** [review]
Super stoked that this is coming together, and looking really good already!

On first round of review, no big changes to functionality. I'll go ahead and make the simpler changes I requested and spend some time with the unit tests tomorrow.

File: R/conformal_infer.R (hunk 4 above)
Comment (topepo) [reply R/conformal_infer.R:30]: The residuals from the (real) training set, meaning the non-trial values

File: R/conformal_infer.R [hunk 5]
```diff
@@ -0,0 +1,459 @@
...
+  )
+}
+
+get_root <- function(x, ctrl) {
+  if (inherits(x, "try-error")) {
+    rlang::warn("Could not finish the search process due to errors.")
+    return(NA_real_)
+  }
+  if (x$iter == ctrl$max_iter) {
+    rlang::warn("Search did not converge.")
```
Comment (simonpcouch) [inline R/conformal_infer.R:413]: Based on the snapshots for this functionality, it seems like this may duplicate warnings from `stats::uniroot`.

File: R/conformal_infer.R [hunk 6]
```diff
@@ -0,0 +1,459 @@
...
+    try(
+      mgcv::gam(sq ~ s(.pred),
+                data = train_res,
+                family = stats::Gamma(link = "log")),
+      silent = TRUE
+    )
+
+  if (inherits(var_mod, "try-error")) {
+    msg <- c(
+      "!" = "The model to estimate the possible interval length failed with the following message:",
```
Comment (simonpcouch) [inline R/conformal_infer.R:226]:
```suggestion
      "The model to estimate the possible interval length failed with the following message:",
```

My suggestion as-is duplicated the exclamation point.😵‍💫

File: R/conformal_infer.R [hunk 7]
```diff
@@ -0,0 +1,459 @@
...
+  } else {
+    upper <- NA_real_
+  }
+  lower <- x[x$trial <= predicted & x$difference >= 0,]
+  if (nrow(lower) > 0) {
+    lower <- max(lower$trial)
+  } else {
+    lower <- NA_real_
+  }
+  dplyr::tibble(.pred_lower = lower, .pred_upper = upper)
```
Comment (simonpcouch) [inline R/conformal_infer.R:351]:
```suggestion
  dplyr::tibble(.pred_lower = lower, .pred = predicted, .pred_upper = upper)
```

This function returns columns `.pred_lower`, `.pred`, and `.pred_upper` on failure, but only `.pred_lower` and `.pred_upper` on success. For type-stability:

File: R/conformal_infer.R [hunk 8]
```diff
@@ -0,0 +1,459 @@
...
+    ctrl,
+    .progress = ctrl$progress,
+    .options = furrr::furrr_options(seed = ctrl$seed, stdout = FALSE)
+  )
+}
+
+grid_one <- function(new_data, model, train_data, level, ctrl) {
+  y_name <- get_outcome_name(model)
+
+  pkg_inst <- lapply(ctrl$required_pkgs, rlang::is_installed)
```
Comment (simonpcouch) [inline R/conformal_infer.R:307]:
```suggestion
  pkg_inst <- lapply(ctrl$required_pkgs, rlang::check_installed)
```

`rlang::is_installed` just returns `TRUE` or `FALSE`--I believe we instead want `rlang::check_installed` here.

File: R/conformal_infer.R [hunk 9]
```diff
@@ -0,0 +1,459 @@
...
+
+get_diff <- function(val, ...) {
+  res <- trial_fit(val, ...)
+  res$difference
+}
+
+optimize_one <- function(new_data, model, train_data, level, ctrl) {
+  y_name <- get_outcome_name(model)
+
+  pkg_inst <- lapply(ctrl$required_pkgs, rlang::is_installed)
```
Comment (simonpcouch) [inline R/conformal_infer.R:378]:
```suggestion
  pkg_inst <- lapply(ctrl$required_pkgs, rlang::check_installed)
```

Same note on `rlang::is_installed()` vs. `rlang::check_installed()`.

**Review by simonpcouch — COMMENTED** [review]
A few more notes after spending some time with the tests. I'll work through the more straightforward changes myself and then drop a note here when I'm wrapped up!

File: R/conformal_infer.R (hunk 8 above)
Comment (topepo) [reply R/conformal_infer.R:307]: My rationale was that we should stop execution if a package is missing.

Comment (simonpcouch) [reply R/conformal_infer.R:307]: Okay, great! `check_installed()` will do that for us.

File: R/conformal_infer.R [hunk 10]
```diff
@@ -0,0 +1,459 @@
...
+    }
+    list(
+      method = method,
+      trial_points = trial_points,
+      var_multiplier = var_multiplier,
+      max_iter = max_iter,
+      tolerance = tolerance,
+      progress = progress,
+      seed = seed
+    )
```
Comment (simonpcouch) [inline R/conformal_infer.R:458]: Currently a bug here where the `required_pkgs` input is not included in output.

File: R/conformal_infer.R [hunk 11]
```diff
@@ -0,0 +1,459 @@
...
+#' required.
+#' @param seed A single integer used to control randomness when models are
+#' (re)fit.
+#' @return A list object with the options given by the user.
+#' @export
+control_conformal_infer <-
+  function(method = "search", trial_points = 100, var_multiplier = 10,
+           max_iter = 100, tolerance = .Machine$double.eps^0.25, progress = FALSE,
+           required_pkgs = character(0), seed = sample.int(10^5, 1)) {
+    method <- rlang::arg_match0(method, c("search", "grid"))
```
Comment (simonpcouch) [inline R/conformal_infer.R:447]: One last comment (I promise):

Is `"search"` the word we want to use for this argument value? `"grid"` refers to a grid _search_ as well. Would `"root"` or some other word be more evocative?

Comment (topepo) [reply R/conformal_infer.R:447]: Good point. I'll change it to "iterative" since that is consistent with our tuning language.
